Phased Builds DIFF (do not merge)

elliot-nelson opened the pull request:
### Demo

Example PR showing the changes made to enable phased builds.

The `main` branch is buildable with current rush (`rush install && rush build`).  It has some setup (for example, it adds the `"_phase:x"` command definitions to the `package.json` files), but doesn't include anything that would be invalid configuration for current rush.

This `phased` branch is not buildable with current rush, but in theory, it is buildable with a rushstack branch that implements the spec described in https://github.com/microsoft/rushstack/issues/2300.

octogonz commented on rigs/acme-classic-rig/profiles/library/config/rush-project.json:
```diff
@@ -1,3 +1,14 @@
+    "_phase:lint": ["temp/lint"],
+    "_phase:test": ["temp/test"],
+    "_phase:readme_updater": ["temp/readme_updater"],
+    "_phase:push_notes": ["temp/push_notes"]
```
Config files are more intuitive if:
- Predefined structure goes in JSON object keys  (`"projectOutputFolderNames"`,  `"associatedPhases"`, etc)
- The user's data goes in JSON values (`"temp/lint"`, `"_phase:compile"`, etc)

Because the above design doesn't follow that convention, a newcomer may wonder whether `"_phase:push_notes"` is a built-in feature of Rush, that they should search for in our docs -- or is it just example data?  It is example data.  But even if we know that, the role is still unclear:  Is `"_phase:push_notes"` the project output folder name?  No, `"temp/push_notes"` seems like a folder name, so then is `"_phase:push_notes"` the corresponding "project" name?  No.  It's a "phase name," but that terminology doesn't appear anywhere in the file format.

If we instead move the data into JSON values, we're forced to introduce a JSON key name that clears up these questions.  Here's a possibility:
```js
{
  "phaseOptions": [
    {
      "phaseName": "_phase:compile",
      "outputFolderNames": ["lib", "lib-commonjs", "dist", "temp/compile"]
    },
    {
      "phaseName": "_phase:lint",
      "outputFolderNames": ["temp/lint"]
    },
    {
      "phaseName": "_phase:test",
      "outputFolderNames": ["temp/test"]
    },
    {
      "phaseName": "_phase:readme_updater",
      "outputFolderNames": ["temp/readme_updater"]
    },
    {
      "phaseName": "_phase:push_notes",
      "outputFolderNames": ["temp/push_notes"]
    }
  ]
}
```


People will be quick to point out some downsides of this approach:
- The person writing the config file has to write more characters
- If the same `"phaseName"` appears twice, that won't be caught for free as a syntax error (`JSON object has a duplicate key "_phase:push_notes"`).  Instead we have to implement custom validation with an arguably more intelligible message (`The phase name "_phase:push_notes" is defined twice in "phaseOptions"`)

But these are concerns of someone who is implementing the system.  A higher priority is perhaps the ***end user***, the newcomer who is trying to figure out how our stuff works, who is making a decision whether to adopt it.

`</personal-soapbox>`
`</grain-of-salt>`

elliot-nelson replied:
That seems reasonable -- will noodle on your suggestion.

octogonz commented on common/config/rush/command-line.json:
```diff
@@ -180,7 +228,15 @@
       // For non-phased commands, we will continue to use the "associatedCommands" setting to associate
       // parameters with them.  (Phased commands must NOT appear in this list.)
-      "associatedCommands": ["build"]
+      "associatedCommands": ["build"],
```
If this is a realistic example, we should show the intended Rush shell commands for realistic everyday tasks:

- How do I build everything including all phases?
- How do I update the README and do nothing else?
- How do I do a `--lite` build where we skip unit tests and linting?
- For these things, what's the equivalent of `rush rebuild`?

Maybe the answer is that initially each of these problems requires a completely new command action name.  (`rush build-lite`?)  But if so we can still enumerate them to get a sense of how painful that feature cut will be in practice.

octogonz commented on common/config/rush/command-line.json:
```diff
@@ -168,6 +168,54 @@
   ],
 
+  /* PHASES */
+  "phases": [
```
How are these phases mapped to Rush actions?  I think we're missing the definition in `"commands"`.

octogonz replied:
If this is a realistic example, we should show the intended Rush shell commands for realistic everyday tasks:

- How do I build everything including all phases?
- How do I update the README and do nothing else?
- How do I do a `--lite` build where we skip unit tests and linting?
- For these things, what's the equivalent of `rush rebuild`?

Maybe the answer is that initially each of these problems requires a completely new command action name.  (`rush build-lite`?)  But if so we can still enumerate them to get a sense of how painful that feature cut will be in practice.

elliot-nelson replied:
These are great questions.

Since presumably the monorepo had this `--lite` need before phases, I implemented a solution in `main` (a custom build script that lets it properly handle `--lite` from the rushx command line).

That `--lite` flag is rough for the phased approach though.  As an example, I imagined that the user wanted `build.js` to pass `--lite` both to Heft and to the `push_notes.js` custom script.  If I invent a new phased command `build-lite`, then it can have phases `['_phase:compile', '_phase:readme_updater']` etc., and you don't need to pass `--lite` to Heft because you've baked that into the phase. But that's not generically possible if you were doing something with the `--lite` flag.

I guess one possible approach is to implement lite _phases_, `push-notes` and `push-notes-lite`, and only include the appropriate one.  But they'll never share build cache, and would have to be leaf nodes (no dependents), otherwise how would a dependent phase know which one to rely on.

octogonz commented on common/config/rush/command-line.json:
```diff
@@ -180,7 +228,15 @@
+      // For example, "rush build" will accept the union of all parameters associated with any of
+      // its individual phases.
+      //
+      // The "--production" flag will be appended to package.json scripts for these specific phases,
```
The `"node scripts/readme_updater.js"` script could print this CLI parameter value to confirm that it was passed along correctly.

elliot-nelson commented:
@octogonz OK, this should be ready for another pass.

Rough summary:

- I've gone beyond the scope of the spec and added proposed properties `addPhasesToCommand` and `skipPhasesForCommand` to the parameter definition.  Perhaps the details need more work, but I feel strongly that this _functionality_ needs to be part of the initial MVP.  Working around the lack of this functionality is too much of a hack to ask people onboarding to phases, I think.

- Added your proposed structure to the `rush-project.json` file.  This was a good suggestion because among other things, I forgot that _other_ bulk commands (which can still use the build cache) still need a place to get their non-phased output folders from.

- Minor tweaking of build script names, contents, and phase names to make them more consistent.
